### PR TITLE
Add support for parameterized queries

### PR DESCRIPTION
**Pull request by shubhanan-ox**

### CODE REVIEW

File: ox_bqpipeline/bqpipeline.py [hunk 1]
```diff
@@ -297,9 +356,66 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+                if not (depth or isinstance(key, str)):
+                    return False
+                value = query_params[key]
+
+            # For list, make sure each element is a scalar type.
+            # For dict, check recursively for valid parameters.
+            # for others, check that they map to valid scalar types.
+            if isinstance(value, list):
+                typ = set([type(val).__name__ for val in value])
+                res = not typ or (len(typ) == 1 and
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: what is `not typ` handling?

File: ox_bqpipeline/bqpipeline.py [hunk 2]
```diff
@@ -297,9 +356,66 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+                                  typ.pop() in bq_scalar_type_map)
+            elif isinstance(value, dict):
+                res = self.validate_query_params(value, depth=1)
+            else:
+                res = type(value).__name__ in bq_scalar_type_map
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: don't use `__name__` (see suggested refactoring of bq_scalar_type_map)

File: ox_bqpipeline/bqpipeline.py [hunk 3]
```diff
@@ -297,9 +356,66 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+
+        :param query_params dict|list: Dict is provided for named parameters,
+                                       otherwise, a list.
+        :return: concrete subtype of bigquery._AbstractQueryParameter
+        """
+        if not query_params:
+            return None
+
+        if not self.validate_query_params(query_params):
+            self.logger.warning('Invalid query parameters provided!')
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: Seems like this should throw a `ValueError`?

File: ox_bqpipeline/bqpipeline.py [hunk 4]
```diff
@@ -297,9 +356,66 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+            else:
+                res = type(value).__name__ in bq_scalar_type_map
+            if not res:
+                return res
+        return True
+
+    def set_query_params(self, query_params):
+        """Set the query parameters
+
+        :param query_params dict|list: Dict is provided for named parameters,
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: When would this be anything but dict?

File: ox_bqpipeline/bqpipeline.py [hunk 5]
```diff
@@ -324,8 +440,13 @@ def run_query(self, path, batch=False, wait=True, create=True,
         template = self.jinja2.from_string(template_str)
         query = template.render(**kwargs)
         client = self.get_client()
+        job_config = self.create_job_config(batch, dest, create, overwrite)
+        query_parameters = self.set_query_params(query_params)
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: nit avoid unnecessary `self.set_query_params` call when `query_params is None` which is true by default.
Also can simplify to 
```python
       if query_params:
           job_config.query_parameters = self.set_query_params(query_params)
```

Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: On second thought, `query_params` should be added as an arg to `create_job_config` and this logic moved there.

File: tests/test_bqpipeline.py [hunk 6]
```diff
@@ -12,16 +12,28 @@
...
+
+from ox_bqpipeline import bqpipeline
+
+
+def mock_gcs_export():
+    pass
 
 
 class TestQueryParameters(unittest.TestCase):
+    def setUp(self):
```
Comment (jaketf) [inline tests/test_bqpipeline.py]: IMO noop setUp can be omitted.

File: tests/test_bqpipeline.py [hunk 7]
```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
...
+        result = bqpipeline.set_scalar_parameter('test', 1, typ='int')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'INT64', 1))
+
+        dt = datetime.datetime.now()
+        result = bqpipeline.set_scalar_parameter('test', dt, typ='datetime')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'TIMESTAMP', dt))
+
+        b = bytes(123)
```
Comment (jaketf) [inline tests/test_bqpipeline.py]: Tests don't cover [all BQ types supported as scalar parameters.](https://googleapis.github.io/google-cloud-python/latest/bigquery/generated/google.cloud.bigquery.query.ScalarQueryParameter.html#google.cloud.bigquery.query.ScalarQueryParameter)

Let's make sure we're covering float bool and date
```
‘STRING’, ‘INT64’, ‘FLOAT64’, ‘NUMERIC’, ‘BOOL’, ‘TIMESTAMP’, ‘DATETIME’, or ‘DATE’.
```

File: ox_bqpipeline/bqpipeline.py [hunk 8]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+def set_list_parameter(key, value, typ):
+    """Set a named array query parameter
+
+    :param key str: key for the named parameter
+    :param value list(int, float, str, byte, datetime.datetime):
+        value for the named parameter
+    :param typ str: type of the array parameter.
+    :return: bigquery.ArrayQueryParameter
+    """
+    typ = type(value[0]).__name__
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: nit: better error handling and earlier type checking 
```suggestion
    if len(value) == 0:
        raise ValueError("BigQuery Array Parameter must be a non-empty list")
    if  len(set(map(type, value))) > 1:
        raise ValueError("All elements of BigQuery Array Parameter should be of the same type")
```

File: tests/test_bqpipeline.py [hunk 9]
```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
...
+    def test_array_parameters(self):
+        result = bqpipeline.set_list_parameter('test', ['abc'], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter('test', 'STRING',
+                                                              ['abc']))
+
+        result = bqpipeline.set_list_parameter('test', [1, 2], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter('test', 'INT64',
+                                                              [1, 2]))
+
+        with self.assertRaises(IndexError):
```
Comment (jaketf) [inline tests/test_bqpipeline.py]: See suggested refactoring in function def. This should raise `ValueError`
Also add test for a list with different types ie `[1, 'two', False]`.

File: tests/test_bqpipeline.py [hunk 10]
```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
...
+            bqpipeline.set_list_parameter('test', [], typ='list')
+
+    def test_struct_parameters(self):
+        result = bqpipeline.set_dict_parameter('test', {'a': 'abc'},
+                                               typ='list')
+        self.assertEqual(result, bigquery.StructQueryParameter(
```
Comment (jaketf) [inline tests/test_bqpipeline.py]: Add test for empty dict. Should we error on empty dict?

File: tests/test_bqpipeline.py [hunk 11]
```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
...
+        result = bqpipeline.set_scalar_parameter(None, 1, typ='int')
+        self.assertEqual(result, bigquery.ScalarQueryParameter(None,
+                                                               'INT64', 1))
+
+        result = bqpipeline.set_list_parameter(None, ['abc'], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter(None, 'STRING',
+                                                              ['abc']))
+
+    def test_query_params(self):
+        bq = bqpipeline.BQPipeline(
```
Comment (jaketf) [inline tests/test_bqpipeline.py]: nit: s/bq/bqp/

File: tests/test_bqpipeline.py [hunk 12]
```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
...
+            job_name='testjob', default_project='testproject',
+            default_dataset='testdataset')
+        cfg = bq.create_job_config(batch=False, create=False, overwrite=False)
+        self.assertIsNone(bq.set_query_params([]))
```
Comment (jaketf) [inline tests/test_bqpipeline.py]: See note above on having `create_job_config` accept a `query_params` arg.
We should refactor this behavior and then have a test of `create_job_config(...,query_params)` and validate it creates the whole job_config as expected for no query params and some query params.

File: tests/test_bqpipeline.py [hunk 13]
```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
...
+                         [bigquery.ScalarQueryParameter('test', 'INT64', 1)])
+
+        # Test invalid query parameters.
+        self.assertIsNone(bq.set_query_params([1, [1, 'two']]))
+
+    def test_invalid_query_params(self):
+        bq = bqpipeline.BQPipeline(
+            job_name='testjob', default_project='testproject',
+            default_dataset='testdataset')
+        cfg = bq.create_job_config(batch=False, create=False, overwrite=False)
```
Comment (jaketf) [inline tests/test_bqpipeline.py]: why are we even making this job config? appears to be unused.

File: ox_bqpipeline/bqpipeline.py [hunk 14]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+
+    :param key str: key for the named parameter
+    :param value (int, float, str, byte, datetime.datetime):
+        value for the named parameter
+    :param typ str: type of the scalar parameter.
+    :return: bigquery.ScalarQueryParameter
+    """
+    return bigquery.ScalarQueryParameter(key, bq_scalar_type_map[typ], value)
+
+def set_list_parameter(key, value, typ):
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 15]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+        value for the named parameter
+    :param typ str: type of the array parameter.
+    :return: bigquery.ArrayQueryParameter
+    """
+    typ = type(value[0]).__name__
+    return bigquery.ArrayQueryParameter(key,
+                                        bq_scalar_type_map[typ],
+                                        value)
+
+def set_dict_parameter(key, value, typ):
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py (hunk 4 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: named parameters will be passed as dict and positional parameters will be passed as a list. Updated the docstring to call that out.

File: ox_bqpipeline/bqpipeline.py (hunk 3 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py (hunk 2 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 16]
```diff
@@ -297,9 +356,66 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+                    return False
+                value = query_params[key]
+
+            # For list, make sure each element is a scalar type.
+            # For dict, check recursively for valid parameters.
+            # for others, check that they map to valid scalar types.
+            if isinstance(value, list):
+                typ = set([type(val).__name__ for val in value])
+                res = not typ or (len(typ) == 1 and
+                                  typ.pop() in bq_scalar_type_map)
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 17]
```diff
@@ -297,9 +356,66 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+            else:
+                if not (depth or isinstance(key, str)):
+                    return False
+                value = query_params[key]
+
+            # For list, make sure each element is a scalar type.
+            # For dict, check recursively for valid parameters.
+            # for others, check that they map to valid scalar types.
+            if isinstance(value, list):
+                typ = set([type(val).__name__ for val in value])
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py (hunk 1 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: As discussed, this will not be needed as empty list will raise a ValueError upfront.

One issue that may pop up is for a use case where default value of empty list is needed for a query. Since the current approach infers the type of the QueryArrayParameter from the contents of the python list, it would not be possible to have empty lists unless an additional parameter is added to specify the BQ type in case the array is empty. I left it out to keep the parameter definition simple. Let me know if we need to reconsider this.

File: tests/test_bqpipeline.py (hunk 10 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: For an empty dict or a list, the use case I was thinking about was when the query is trying to set the default value to empty dict or list respectively. In case of list, it is not possible to set such parameters without the knowledge of the array type - we currently infer that from the contents of the python list. In case of a dict, however, that problem does not exist as the struct can be set with empty contents without having to infer any type info for the contents. Let me know your thoughts on the same.

File: ox_bqpipeline/bqpipeline.py (hunk 8 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: tests/test_bqpipeline.py (hunk 6 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: Done.

File: tests/test_bqpipeline.py (hunk 12 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: Done. Also, note the change in method invocation. Named keyword parameters should always be passed as such otherwise, they may get wrong value unintentionally.

File: ox_bqpipeline/bqpipeline.py (hunk 5 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 18]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+def set_dict_parameter(key, value, typ):
+    """Set a named struct query parameter
+
+    :param key str: key for the named parameter
+    :param value dict: value for the named parameter
+    :param typ str: type of the scalar parameter.
+    :return: bigquery.ScalarQueryParameter
+    """
+    res = []
+    for k in value:
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 19]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+    return bigquery.ArrayQueryParameter(key,
+                                        bq_scalar_type_map[typ],
+                                        value)
+
+def set_dict_parameter(key, value, typ):
+    """Set a named struct query parameter
+
+    :param key str: key for the named parameter
+    :param value dict: value for the named parameter
+    :param typ str: type of the scalar parameter.
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 20]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+
+    :param key str: key for the named parameter
+    :param value list(int, float, str, byte, datetime.datetime):
+        value for the named parameter
+    :param typ str: type of the array parameter.
+    :return: bigquery.ArrayQueryParameter
+    """
+    typ = type(value[0]).__name__
+    return bigquery.ArrayQueryParameter(key,
+                                        bq_scalar_type_map[typ],
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 21]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+def set_scalar_parameter(key, value, typ):
+    """Set a named scalar query parameter
+
+    :param key str: key for the named parameter
+    :param value (int, float, str, byte, datetime.datetime):
+        value for the named parameter
+    :param typ str: type of the scalar parameter.
+    :return: bigquery.ScalarQueryParameter
+    """
+    return bigquery.ScalarQueryParameter(key, bq_scalar_type_map[typ], value)
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 22]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
                         job.job_id)
     return wrapper
 
+def set_scalar_parameter(key, value, typ):
+    """Set a named scalar query parameter
+
+    :param key str: key for the named parameter
+    :param value (int, float, str, byte, datetime.datetime):
+        value for the named parameter
+    :param typ str: type of the scalar parameter.
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 23]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+    """
+    return bigquery.ScalarQueryParameter(key, bq_scalar_type_map[typ], value)
+
+def set_list_parameter(key, value, typ):
+    """Set a named array query parameter
+
+    :param key str: key for the named parameter
+    :param value list(int, float, str, byte, datetime.datetime):
+        value for the named parameter
+    :param typ str: type of the array parameter.
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 24]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+    :param typ str: type of the scalar parameter.
+    :return: bigquery.ScalarQueryParameter
+    """
+    res = []
+    for k in value:
+        typ = type(value[k]).__name__
+        res.append(set_parameter[typ](k, value[k], typ))
+    return bigquery.StructQueryParameter(key, *res)
+
+set_parameter = {
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 25]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
                         job.job_id)
     return wrapper
 
+def set_scalar_parameter(key, value, typ):
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 26]
```diff
@@ -141,6 +142,66 @@ def wrapper(*args, **kwargs):
...
+    'int'        : set_scalar_parameter,
+    'float'      : set_scalar_parameter,
+    'bytes'      : set_scalar_parameter,
+    'datetime'   : set_scalar_parameter,
+    'list'       : set_list_parameter,
+    'dict'       : set_dict_parameter,
+}
+
+bq_scalar_type_map = {
+    'str'   : 'STRING',
```
[the remark this answers is not in the document]
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: tests/test_bqpipeline.py (hunk 11 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: Done.

**Comment by shubhanan-ox** [comment]
Still need to address a small few comments and add more unit tests around the type dict mapping. However, major refactoring stuff has been addressed and ready for a look. Also, had a couple of questions regarding the empty parameters handling. PTAL and let me know. Thanks!

File: tests/test_bqpipeline.py (hunk 7 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: Done.

File: tests/test_bqpipeline.py (hunk 9 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: This is done in the test_query_params unit test. I've added additional unit tests with more scenarios for other types.

File: tests/test_bqpipeline.py (hunk 13 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py [hunk 27]
```diff
@@ -32,6 +32,20 @@
...
+    str   : 'STRING',
+    int   : 'INT64',
+    float  : 'FLOAT64',
+    bool: 'BOOL',
+    bytes  : 'BYTES',
+    datetime.datetime: 'TIMESTAMP',
+    datetime.date: 'DATE',
+}
+
+NUMERIC_BOUNDS = [-99999999999999999999999999999.999999999,
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: nit: for clarity refactor this to `min max` keyed dict
```python 
NUMERIC_BOUNDS = {
    'min': -99999999999999999999999999999.999999999,
    'max': 99999999999999999999999999999.999999999
}
```

File: ox_bqpipeline/bqpipeline.py [hunk 28]
```diff
@@ -142,64 +156,41 @@ def wrapper(*args, **kwargs):
...
-    'bytes'  : 'BYTES',
-    'datetime': 'TIMESTAMP',
-    'date': 'DATE',
-}
+    if (isinstance(value, (str, int, bytes, bool, datetime.datetime,
+                           datetime.date))):
+        return bigquery.ScalarQueryParameter(
+            key, BQ_SCALAR_TYPE_MAP.get(type(value)), value)
+    elif isinstance(value, float):
+        if value < NUMERIC_BOUNDS[0] or value > NUMERIC_BOUNDS[1]:
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: nit: see note on refactoring to make this a dict w/ keys min and max.

File: ox_bqpipeline/bqpipeline.py [hunk 29]
```diff
@@ -350,72 +341,82 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+        else:
+            datatype = set(map(type, query_params))
+            # Make sure all keys are str
+            res = len(datatype) == 1 and datatype.issubset([str])
+            res = res and all([self.validate_parameter(query_params[key])
+                               for key in query_params])
+        return res
+
+    def validate_parameter(self, query_parameter):
+        """Validate individual query parameter."""
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: add to docstring the class (or super class) of query_parameter.

File: ox_bqpipeline/bqpipeline.py [hunk 30]
```diff
@@ -350,72 +341,82 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+        res = True
+        if isinstance(query_parameter, list):
             # For list, make sure each element is a scalar type.
+            datatype = set(map(type, query_parameter))
+            res = (len(datatype) == 1 and
+                   datatype.issubset(BQ_SCALAR_TYPE_MAP))
+        elif isinstance(query_parameter, dict):
+            if not query_parameter:
+                return True
+            # For dict, make sure all keys are same type and
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: Do all keys need to be the same type for dict parameters? this seems like an odd constraint.

File: ox_bqpipeline/bqpipeline.py [hunk 31]
```diff
@@ -350,72 +341,82 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+            # one of BQ supported types.
+            datatype = set(map(type, query_parameter))
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: This is gonna check the type of the keys in query_parameter. Which is likely just going to be all string.

File: ox_bqpipeline/bqpipeline.py [hunk 32]
```diff
@@ -350,72 +341,82 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
+            datatype = set(map(type, query_params))
+            # Make sure all keys are str
+            res = len(datatype) == 1 and datatype.issubset([str])
+            res = res and all([self.validate_parameter(query_params[key])
+                               for key in query_params])
+        return res
+
+    def validate_parameter(self, query_parameter):
+        """Validate individual query parameter."""
+        res = True
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: nit: more descriptive name
s/res/is_valid_parameter/

File: ox_bqpipeline/bqpipeline.py [hunk 33]
```diff
@@ -350,72 +341,82 @@ def create_job_config(self, batch=False, dest=None, create=True,
...
-            if isinstance(query_params, list):
-                value = key
-                key = None
-            else:
-                if not (depth or isinstance(key, str)):
-                    return False
-                value = query_params[key]
-
+        if isinstance(query_params, list):
+            res = all([self.validate_parameter(value) for value in
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: nit: more descriptive name 
s/res/all_params_valid/

File: tests/test_bqpipeline.py (hunk 10 above)
Comment (jaketf) [reply tests/test_bqpipeline.py]: I disagree about dict.
A struct w/o any keys makes no in BQ. A struct can either be `null` or a struct w/ keys whose values are null. 
I'd be surprised if BQ didn't throw an error for passing a query param of `{}`

File: ox_bqpipeline/bqpipeline.py (hunk 30 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Struct keys could be any scalar type. Not sure why I did this - probably trickled from the named parameters constraint. Removed.

File: ox_bqpipeline/bqpipeline.py (hunk 31 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Yes. Fixed.

File: ox_bqpipeline/bqpipeline.py (hunk 32 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py (hunk 33 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: This would have to be the same variable that's returned at the end of the function. Instead, I have updated to return here itself.

**Comment by shubhanan-ox** [comment]
Thanks for the review! I am working on the pydoc. I am seeing some issues with the rendered markdown so, working through that.

File: tests/test_bqpipeline.py (hunk 10 above)
Comment (shubhanan-ox) [reply tests/test_bqpipeline.py]: Done.

**Comment by shubhanan-ox** [comment]
I've made the changes discussed today. PTAL. Thanks!

File: ox_bqpipeline/bqpipeline.py [hunk 34]
```diff
@@ -422,10 +422,32 @@ def set_query_params(self, query_params):
...
+    def get_query_details(self, query_details):
+        """Returns the query path, destination, parameters and GCS flag.
+
+        :param query_details <tuple|str>:
+        :returns: Query path, destination GCS path or table spec, query params,
+                and boolean to signify if the destination is GCS.
+        """
+        sql_path, destination, query_params = None, None, None
+        is_gcs_dest = False
+        if type(query_details) == tuple:
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: nit: use `isinstance` to check type.

Comment (jaketf) [reply ox_bqpipeline/bqpipeline.py]: move the length check to this if statement.
```suggestion
        if isinstance(query_details, tuple) and len(query_details) > 1:
```

File: ox_bqpipeline/bqpipeline.py [hunk 35]
```diff
@@ -422,10 +422,32 @@ def set_query_params(self, query_params):
...
+            sql_path = query_details[0]
+            destination = query_details[1]
+            if len(query_details) > 1:
```
Comment (jaketf) [inline ox_bqpipeline/bqpipeline.py]: see above note to decrease level of nested ifs and avoid checking len(...) == 3 when len(...) <= 1.

Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

File: ox_bqpipeline/bqpipeline.py (hunk 34 above)
Comment (shubhanan-ox) [reply ox_bqpipeline/bqpipeline.py]: Done.

**Comment by shubhanan-ox** [comment]
Done.

**Comment by shubhanan-ox** [comment]
This addresses #9
